### ISSUE TITLE
Stale Config Settings

### ISSUE THREAD

**Issue by deckar01** [AtomLinter/linter-eslint#892]
## Issue Type
Bug

## Issue Description
Some of the settings in my eslintrc files are not respected after they are changed. I have to completely restart atom to get the correct behavior.

This happens constantly. The most frustrating part is that it is not just ignoring all the rules in the file, just some of them, which creates a time suck of misguided debugging.

## Bug Checklist
- Restarting Atom temporarily resolves the issue
- `eslint` CLI gives the proper result, but `linter-eslint` does not

~I am not going to spend any more time to repro this bug or provide debug info, but it has happened 4-5 times this month. I am switching to [fast-eslint](https://github.com/Adezandee/fast-eslint), which so far seems to be more stable.~

```
Atom version: 1.17.0-dev-4f87dad60
linter-eslint version: 8.1.6
ESLint version: 3.12.2
Hours since last Atom restart: 0.1
Platform: linux
Using local project ESLint from: /home/jared/src/equity_web/node_modules/eslint
Current file's scopes: [
  "source.js",
  "meta.function-call.js",
  "entity.name.function.js"
]
linter-eslint configuration: {
  "lintHtmlFiles": false,
  "useGlobalEslint": false,
  "showRuleIdInMessage": true,
  "disableWhenNoEslintConfig": true,
  "eslintrcPath": "",
  "globalNodePath": "",
  "advancedLocalNodeModules": "",
  "eslintRulesDir": "",
  "disableEslintIgnore": false,
  "disableFSCache": false,
  "fixOnSave": false,
  "scopes": [
    "source.js",
    "source.jsx",
    "source.js.jsx",
    "source.babel",
    "source.js-semantic"
  ],
  "rulesToSilenceWhileTyping": [],
  "rulesToDisableWhileFixing": []
}
```

**Comment by Arcanemagus** [comment]
> I am not going to spend any more time to repro this bug or provide debug info

Unfortunately we have been unable to reproduce this the few other times it has been reported, and as you didn't even post the debug information requested in the template we have _nothing_ to go on here.

> I am switching to fast-eslint

Glad to hear you've found something that works for you!

**Comment by deckar01** [comment]
@Arcanemagus Sorry for being rude. I was frustrated. I added the debug info to the description.

I think I have nailed down some steps that reproduce the issue consistently:
 1. Have a eslintrc file in the root of the project (mine is in `package.json`)
 2. Have a eslintrc file in a child directory (mine is in `spec/.eslintrc`)
 3. Have some undefined variables in a file in the child directory (mine is in`spec/test.js`)
 4. View the file and the errors
 5. Turn off the `no-undef` rule in the child eslintrc
 6. Repoen the file with undefined variables

 - Running eslint from the command line respects the changes, but linter-eslint does not.
 - If I change the env in the child eslintrc file linter-eslint reflects those changes, but still does not apply the modified rules.

**Comment by Arcanemagus** [comment]
Thanks for the steps, I'll try to reproduce that locally and see if I can figure out what's going on.

**Comment by Arcanemagus** [comment]
Unfortunately I'm still unable to reproduce this 😞.

Using these files: [linter-eslint_GH892.zip](https://github.com/AtomLinter/linter-eslint/files/948262/linter-eslint_GH892.zip)

I followed these steps:

1. Run `npm install` in the directory to install the exact version of ESLint used in your debugging information.
2. Open Atom in the directory, and open `foo.js` and note that ESLint shows errors for bone `console` and `bar` being undefined.
3. Uncomment the environment section of the `.eslintrc.js` file in the spec folder and save the file
4. Trigger a lint of `foo.js`, note that the `console` error has disappeared as expected.
5. Uncomment the `rules` section of `.eslintrc.js` and save the file
6. Trigger a lint of `foo.js`, note that the `bar` error has disappeared as expected.

The only difference from the example files here and your description is the use of `.eslintrc.js` vs the old style `.eslintrc`, however I also tested with that style of file with no difference to the results.

Are you seeing anything different using the above files?

**Comment by deckar01** [comment]
I was able to repro with your zip, but I had the most important step out of order.

You have to create the child eslint config file after the child file is linted.

 1. Have an eslintrc file in the root of the project (mine is in `package.json`)
 3. Have some undefined variables in a file in the child directory (mine is in`spec/test.js`)
 4. View the file and the errors
 2. **Create an eslintrc file in a child directory (mine is in `spec/.eslintrc`)**
 5. Turn off the `no-undef` rule in the child eslintrc
 6. Refresh the file with undefined variables

In your example, the browser env gets applied correctly, but the no-undef rule does not. To reproduce, you should just have to temporarily delete the `.eslintrc.js` file, lint `foo.js`, add `.eslintrc.js` (uncomment the config values), and lint `foo.js` again.

Every time I checked out a branch that introduces a new eslint config file with a rule in it, I hit this bug.

**Comment by Arcanemagus** [comment]
Ah ha! That's the critical bit of information that's been missing every other time. I was able to reproduce that.

I suspect this is a bug in ESLint itself since from the CLI a new instance is started for every lint call, while we reuse instances as much as possible to avoid the startup hit whenever we can. I'll do a deeper dig into this later to confirm where the problem is.

**Comment by Arcanemagus** [comment]
Traced this down to our usage of a cached find [here](https://github.com/AtomLinter/linter-eslint/blob/v8.1.6/src/worker-helpers.js#L106). I think we are just going to have to go to the un-cached `find()` to fix this.

**Comment by Arcanemagus** [comment]
@deckar01 There is already an option implemented that will fix this for you: The "Disable FileSystem Cache" option. Enabling that will make this work, issues like this though are making me wonder if we should remove the option or change it's default.

**Comment by deckar01** [comment]
Creating new eslintrc files seems like an important use case. Adding an eslintrc file in branch caused me to hit this bug every time I checked out the branch.

I would recommend disabling the file system cache and just ignore the cache option until a permanent solution is found. I suspect it is possible to detect the config files changing and invalidate the cache, but I tend to prioritize correctness over performance.

**Comment by IanVS** [comment]
Update: I have discussed this with @Arcanemagus and we think we have a solution which does not involve changing any settings.  We seem to be passing a (sometimes cached) config file to ESLint instead of letting it do its own config file resolution.  If we let ESLint handle resolution itself, this bug should be resolved.

**Comment by Arcanemagus** [comment]
Published in v8.1.7 🎉

Thanks for that crucial bit of info leading to tracking this down!

**Comment by deckar01** [comment]
Thanks for fixing this @Arcanemagus.
